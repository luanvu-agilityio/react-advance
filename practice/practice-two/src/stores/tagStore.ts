import { create } from 'zustand'
import type { ProductTag } from '@utils/tagUtils'

interface TagState {
  selectedTags: ProductTag[]
  addTag: (tag: ProductTag) => void
  removeTag: (tag: ProductTag) => void
  clearTags: () => void
  isTagSelected: (tagLabel: string) => boolean
}

export const useProductTagStore = create<TagState>()((set, get) => ({
  selectedTags: [],

  addTag: (tag) => {
    set((state) => {
      if (state.selectedTags.some((t) => t.label === tag.label)) {
        // Tag already exists
        return state
      }
      return { selectedTags: [...state.selectedTags, tag] }
    })
  },

  removeTag: (tag) => {
    set((state) => ({
      selectedTags: state.selectedTags.filter(
        (t) => !(t.label === tag.label && t.type === tag.type)
      ),
    }))
  },

  clearTags: () => set({ selectedTags: [] }),

  isTagSelected: (tagLabel) => {
    return get().selectedTags.some((tag) => tag.label === tagLabel)
  },
}))
