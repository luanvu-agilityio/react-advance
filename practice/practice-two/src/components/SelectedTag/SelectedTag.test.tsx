import { render, fireEvent, screen } from '@testing-library/react'
import SelectedTags from './SelectedTag'
import { useProductTagStore } from '@stores/tagStore'

// Mock the tag store hook
jest.mock('@stores/tagStore', () => ({
  useProductTagStore: jest.fn(),
}))

// SNAPSHOT TESTING
describe('SelectedTags - Rendering', () => {
  beforeEach(() => {
    // Reset the mock before each test
    jest.clearAllMocks()
  })

  test('renders nothing when no tags are selected', () => {
    // Setup the mock to return empty tags
    ;(useProductTagStore as unknown as jest.Mock).mockReturnValue({
      selectedTags: [],
      removeTag: jest.fn(),
    })

    const { container } = render(<SelectedTags />)
    expect(container.firstChild).toBeNull()
  })

  test('renders tags correctly when tags are present', () => {
    // Setup mock with some example tags
    const mockTags = [
      { type: 'brand', label: 'Apple' },
      { type: 'price', label: '$100-$200' },
    ]
    ;(useProductTagStore as unknown as jest.Mock).mockReturnValue({
      selectedTags: mockTags,
      removeTag: jest.fn(),
    })

    const { container } = render(<SelectedTags />)
    expect(container).toMatchSnapshot()
  })

  const mockTags = [
    { type: 'brand', label: 'Apple' },
    { type: 'price', label: '$100-$200' },
    { type: 'category', label: 'Electronics' },
  ]
  ;(useProductTagStore as unknown as jest.Mock).mockReturnValue({
    selectedTags: mockTags,
    removeTag: jest.fn(),
  })

  render(<SelectedTags />)

  // Check header is present
  expect(screen.getByText('Applied filters:')).toBeInTheDocument()

  // Check all tags are rendered
  expect(screen.getByText('Apple')).toBeInTheDocument()
  expect(screen.getByText('$100-$200')).toBeInTheDocument()
  expect(screen.getByText('Electronics')).toBeInTheDocument()
})

// INTERACTIVE BEHAVIOR TESTING
describe('SelectedTags - Interactive', () => {
  test('calls removeTag when remove button is clicked', () => {
    const mockRemoveTag = jest.fn()
    const mockTags = [
      { type: 'brand', label: 'Apple' },
      { type: 'price', label: '$100-$200' },
    ]
    ;(useProductTagStore as unknown as jest.Mock).mockReturnValue({
      selectedTags: mockTags,
      removeTag: mockRemoveTag,
    })

    render(<SelectedTags />)

    // Click the remove button of the first tag
    const removeButtons = screen.getAllByRole('button')
    fireEvent.click(removeButtons[0])

    // Check that removeTag was called with the correct tag
    expect(mockRemoveTag).toHaveBeenCalledWith(mockTags[0])
  })

  test('calls removeTag with correct tag when multiple tags exist', () => {
    const mockRemoveTag = jest.fn()
    const mockTags = [
      { type: 'brand', label: 'Apple' },
      { type: 'price', label: '$100-$200' },
      { type: 'category', label: 'Electronics' },
    ]
    ;(useProductTagStore as unknown as jest.Mock).mockReturnValue({
      selectedTags: mockTags,
      removeTag: mockRemoveTag,
    })

    render(<SelectedTags />)

    // Find all remove buttons
    const removeButtons = screen.getAllByRole('button')

    // Click the second tag's remove button
    fireEvent.click(removeButtons[1])

    // Check that removeTag was called with the second tag
    expect(mockRemoveTag).toHaveBeenCalledWith(mockTags[1])

    expect(mockRemoveTag).not.toHaveBeenCalledWith(mockTags[0])
    expect(mockRemoveTag).not.toHaveBeenCalledWith(mockTags[2])
  })
})
