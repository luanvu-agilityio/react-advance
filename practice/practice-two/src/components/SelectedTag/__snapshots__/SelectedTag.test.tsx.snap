// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedTags - Rendering renders tags correctly when tags are present 1`] = `
<div>
  <div
    class="sc-Qotzb dGhYAe"
  >
    <div
      class="sc-fYsHOw kkBnGR"
    >
      Applied filters:
    </div>
    <div
      class="sc-dIMoHT haFOUc"
    >
      <div
        class="sc-dKREkF dtLIgC"
      >
        Apple
        <button
          class="sc-fWnslK dpwZYk"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-x"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
      </div>
      <div
        class="sc-dKREkF dtLIgC"
      >
        $100-$200
        <button
          class="sc-fWnslK dpwZYk"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-x"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
